pkg/agent/openflow/multicluster.go

```diff
@@ -0,0 +1,164 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package openflow
+
+import (
+	"net"
+
+	"antrea.io/antrea/pkg/agent/openflow/cookie"
+	binding "antrea.io/antrea/pkg/ovs/openflow"
+)
+
+// GlobalVirtualMACForMulticluster is a vritual MAC which will be used only
+// for cross-cluster traffic to distinguish from in-cluster traffic.
+var GlobalVirtualMACForMulticluster, _ = net.ParseMAC("aa:bb:cc:dd:ee:f0")
+
+type featureMulticluster struct {
+	cookieAllocator cookie.Allocator
+	cachedFlows     *flowCategoryCache
+	category        cookie.Category
+	ipProtocols     []binding.Protocol
+	dnatCtZones     map[binding.Protocol]int
+	snatCtZones     map[binding.Protocol]int
+}
+
+func (f *featureMulticluster) getFeatureName() string {
+	return "Multicluster"
+}
+
+func newFeatureMulticluster(cookieAllocator cookie.Allocator, ipProtocols []binding.Protocol) *featureMulticluster {
+	snatCtZones := make(map[binding.Protocol]int)
+	dnatCtZones := make(map[binding.Protocol]int)
+	snatCtZones[ipProtocols[0]] = SNATCtZone
+	dnatCtZones[ipProtocols[0]] = CtZone
+	return &featureMulticluster{
+		cookieAllocator: cookieAllocator,
+		cachedFlows:     newFlowCategoryCache(),
+		category:        cookie.Multicluster,
+		ipProtocols:     ipProtocols,
+		snatCtZones:     snatCtZones,
+		dnatCtZones:     dnatCtZones,
+	}
+}
+
+func (f *featureMulticluster) initFlows() []binding.Flow {
+	return []binding.Flow{}
+}
+
+func (f *featureMulticluster) replayFlows() []binding.Flow {
+	return getCachedFlows(f.cachedFlows)
+}
+
+func (f *featureMulticluster) l3FwdFlowToRemoteViaTun(
+	localGatewayMAC net.HardwareAddr,
+	dstMac net.HardwareAddr,
+	peerServiceCIDR net.IPNet,
+	tunnelPeer net.IP,
+	remoteGatewayIP net.IP) []binding.Flow {
+	ipProtocol := getIPProtocol(peerServiceCIDR.IP)
+	cookieID := f.cookieAllocator.Request(f.category).Raw()
+	var flows []binding.Flow
+	flows = append(flows,
+		// This generates the flow to forward cross-cluster request packets based
+		// on Service ClusterIP range.
+		L3ForwardingTable.ofTable.BuildFlow(priorityNormal).
+			Cookie(cookieID).
+			MatchProtocol(ipProtocol).
+			MatchDstIPNet(peerServiceCIDR).
+			Action().SetSrcMAC(localGatewayMAC). // Rewrite src MAC to local gateway MAC.
+			Action().SetDstMAC(dstMac).          // Rewrite dst MAC to virtual MAC.
+			Action().SetTunnelDst(tunnelPeer).   // Flow based tunnel. Set tunnel destination.
+			Action().LoadRegMark(ToTunnelRegMark).
+			Action().GotoTable(L3DecTTLTable.GetID()).
+			Done(),
+		// This generates the flow to forward cross-cluster reply traffic based
+		// on Gateway IP.
+		L3ForwardingTable.ofTable.BuildFlow(priorityNormal).
+			Cookie(cookieID).
+			MatchProtocol(ipProtocol).
+			MatchCTStateRpl(true).
+			MatchCTStateTrk(true).
+			MatchDstIP(remoteGatewayIP).
+			Action().SetSrcMAC(localGatewayMAC).
+			Action().SetDstMAC(GlobalVirtualMACForMulticluster).
```

Comment:
Not matter it's in-cluster or out-cluster traffic, `dstMac` will be `GlobalVirtualMAC` for request packets from regular Node.

Reply:
I don't mean parameterization of the second mac but don't understand why request and reply use different MAC on regular Node.
If I understand correctly, according to the two flows:
* a request from regular Node to Gateway Node uses `dstMac` which is set to `GlobalVirtualMAC`
* a reply from regular Node to Gateway Node uses `GlobalVirtualMACForMulticluster`

Reply:
yes, but these two flows are not pair, the second reply is for the traffic when the request is initialized from remote cluster which has been SNAT to remote GatewayIP.

Reply:
They are in same data path and same direction and share same next hop, why they use different MAC? what we want `GlobalVirtualMACForMulticluster` stand for?

Reply:
It's mainly used in Gateway flow to distinguish it's cross-cluster or in-cluster traffic in Classifier table, I set the `dstMac` in regular Node as `GlobalVirtualMAC`  considering it's still between Nodes in the same cluster, but I think it can be `GlobalVirtualMACForMulticluster` to have symmetry flows. I will check.

Reply:
Removed the `dstMac`, all is using `GlobalVirtualMACForMulticluster` now.